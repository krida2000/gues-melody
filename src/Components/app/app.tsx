import {PureComponent} from "react";
import * as React from "react";
import WelcomeScreen from '../welcome-screen/welcome-screen';
import LoseScreen from "../lose-screen/lose-screen";
import ArtistQuestionScreen from '../artist-question-screen/artist-question-screen';
import GenreQuestionScreen from '../genre-question-screen/genre-question-screen';
import withActivePlayer from "../../hocs/with-active-player/with-active-player";
import withUserAnswer from "../../hocs/with-user-answer/with-user-answer";
import AuthorizationScreen from "../authorization-screen/authorization-screen";
import {connect} from "react-redux";
import {ActionCreator as GameActionCreator} from "../../reducer/game/game"
import {Operation as DataOperation} from "../../reducer/data/data"
import {Operation as UserOperation} from "../../reducer/user/user"
import {Switch, Route, Redirect, BrowserRouter} from "react-router-dom";
import {appProps, artistUserAnswer, genreUserAnswer} from "../../types";
import history from '../../history'
import WinScreen from "../win-screen/win-screen";

class App extends PureComponent<appProps, {}> {
  _getScreen() {

    const {questions, mistakes, maxMistakes, onUserAnswer, step, resetGame, isAuthorizationRequired} = this.props;
    const question = questions[step];

    if(mistakes > maxMistakes){
      return <Redirect to={`/lose`} />;
    }

    if (isAuthorizationRequired) {
      return <Redirect to={`/auth`} />;
    }

    if (step === -1) {
      const {onWelcomeButtonClick} = this.props;

      return <WelcomeScreen
        errorsCount={maxMistakes}
        onWelcomeButtonClick={() => onWelcomeButtonClick()}
      />;
    }

    if (step === 10) {
      return <Redirect to={`/win`} />;
    }

    if (questions[step] && questions[step].type === `artist`) {
      const ArtistQuestionScreenWithActivePlayer = withActivePlayer(ArtistQuestionScreen);

      return <ArtistQuestionScreenWithActivePlayer onAnswer={(userAnswer: artistUserAnswer) => onUserAnswer(userAnswer, question)}
        question={questions[step]}
        mistakes={mistakes}/>;
    }

    if (questions[step] && questions[step].type === `genre`) {
      const GenreQuestionScreenWithActivePlayerAndAnswer = withUserAnswer(withActivePlayer(GenreQuestionScreen));

      return <GenreQuestionScreenWithActivePlayerAndAnswer onAnswer={(userAnswer: genreUserAnswer) => onUserAnswer(userAnswer, question)}
        question={questions[step]}
        mistakes={mistakes}/>;
    }

    resetGame();

    return null;
  }

  render() {
    const {onAuth, resetGame, isAuthorizationRequired, mistakes, maxMistakes, step} = this.props;

    return <BrowserRouter>
      <Switch>
        <Route exact path={`/`} render={() => this._getScreen()}/>
        <Route exact path={`/auth`} render={() => <AuthorizationScreen onAuthorization={onAuth} isAuthorizationRequired={isAuthorizationRequired}/>}/>
        <Route exact path={`/lose`} render={() => <LoseScreen replayButtonClickHandler={resetGame} mistakes={mistakes} maxMistakes={maxMistakes}/>}/>
        <Route exact path={`/win`} render={() => <WinScreen replayButtonClickHandler={resetGame} mistakes={mistakes} maxMistakes={maxMistakes} steep={step}/>}/>
      </Switch>
    </BrowserRouter>;
  }
}

// App.propTypes = {
//   questions: PropTypes.array.isRequired,
//   mistakes: PropTypes.number.isRequired,
//   maxMistakes: PropTypes.number.isRequired,
//   gameTime: PropTypes.number.isRequired,
//   step: PropTypes.number.isRequired,
//   isAuthorizationRequired: PropTypes.bool.isRequired,
//   onUserAnswer: PropTypes.func.isRequired,
//   onWelcomeButtonClick: PropTypes.func.isRequired,
//   resetGame: PropTypes.func.isRequired,
//   onAuth: PropTypes.func.isRequired,
//   // questions: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.shape({
//   //   type: `artist`,
//   //   song: {
//   //     artist: PropTypes.string.isRequired,
//   //     src: PropTypes.string.isRequired
//   //   },
//   //   answers: PropTypes.arrayOf({
//   //     picture: PropTypes.string.isRequired,
//   //     artist: PropTypes.string.isRequired
//   //   }),
//   // }), PropTypes.shape({
//   //   type: `genre`,
//   //   genre: PropTypes.string.isRequired,
//   //   answers: PropTypes.arrayOf({
//   //     src: PropTypes.string.isRequired,
//   //     genre: PropTypes.string.isRequired
//   //   }),
//   // })])).isRequired,
// };

const mapStateToProps = (state, ownProps) => {
  return Object.assign({}, ownProps, {
    step: state.game.step,
    mistakes: state.game.mistakes,
    questions: state.data.questions,
    isAuthorizationRequired: state.user.isAuthorizationRequired,
  });
};

const mapDispatchToProps = (dispatch) => ({
  onWelcomeButtonClick: () => {
    dispatch(GameActionCreator.incrementStep());
  },
  onUserAnswer: (userAnswer, question, mistakes, maxMistakes) => {
    dispatch(GameActionCreator.incrementStep());
    dispatch(GameActionCreator.incrementMistakes(userAnswer, question));
  },
  onAuth: (authData) => {
    dispatch(UserOperation.login(authData));
  },
  resetGame() {
    dispatch(GameActionCreator.reset());
    dispatch(DataOperation.loadQuestions());
  },
});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);
