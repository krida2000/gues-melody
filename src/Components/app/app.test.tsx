import * as React from "react";
import * as renderer from "react-test-renderer";
import {App} from "./app";
const noop = () => {
  // do nothing
};

it(`App screen correctly render`, () => {
  const tree = renderer
    .create(
      <App
        isAuthorizationRequired={false}
        mistakes={0}
        maxMistakes={0}
        onAuth={noop}
        onUserAnswer={noop}
        onWelcomeButtonClick={noop}
        questions={[]} resetGame={noop}
        step={-1}
      />).toJSON();

  expect(tree).toMatchSnapshot();
});
