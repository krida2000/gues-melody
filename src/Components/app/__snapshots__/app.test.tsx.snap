// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App screen correctly render App screen correctly render 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[Function]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      Нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      0
       ошибок.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;

exports[`App screen correctly render App screen with authorization correctly render 1`] = `
<form
  action=""
  className="login__form"
>
  <p
    className="login__field"
  >
    <label
      className="login__label"
      htmlFor="name"
    >
      Логин
    </label>
    <input
      className="login__input"
      id="name"
      name="name"
      type="text"
    />
  </p>
  <p
    className="login__field"
  >
    <label
      className="login__label"
      htmlFor="password"
    >
      Пароль
    </label>
    <input
      className="login__input"
      id="password"
      name="password"
      type="password"
    />
    <span
      className="login__error"
    >
      Неверный пароль
    </span>
  </p>
  <button
    className="login__button button"
    onClick={[Function]}
    type="submit"
  >
    Войти
  </button>
</form>
`;

exports[`App screen correctly render App screen with lose correctly render 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Какая жалость!
  </h2>
  <p
    className="result__total result__total--fail"
  >
    У вас закончились все попытки. Ничего, повезёт в следующий раз!
  </p>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Попробовать ещё раз
  </button>
</section>
`;

exports[`App screen correctly render App screen with win correctly render 1`] = `
<section
  className="result"
>
  <div
    className="result__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="result__title"
  >
    Ура победа!
  </h2>
  <p
    className="result__total result__total--fail"
  >
    Вы ответили на все вопросы, сделав всего 
    0
     шибок из допустимых 
    0
    !
  </p>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Попробовать ещё раз
  </button>
</section>
`;
